src/storage/v3/key_store.cpp

```diff
@@ -0,0 +1,40 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <algorithm>
+#include <iterator>
+#include <ranges>
+
+#include "storage/v3/key_store.hpp"
+#include "storage/v3/property_value.hpp"
+
+namespace memgraph::storage::v3 {
+
+KeyStore::KeyStore(const std::vector<PropertyValue> &key_values) {
+  for (auto i = 0; i < key_values.size(); ++i) {
+    MG_ASSERT(!key_values[i].IsNull());
+    store_.SetProperty(PropertyId::FromInt(i), key_values[i]);
+  }
+}
+
+PropertyValue KeyStore::GetKey(const size_t index) const { return store_.GetProperty(PropertyId::FromUint(index)); }
+
+std::vector<PropertyValue> KeyStore::Keys() const {
+  auto keys_map = store_.Properties();
```

Comment:
Basically, you get a `copy` of the properties here which you then move to the `keys`. Can the `properties` change by the end of this function call (through some other thread which sets the properties)? If that's the case, by the end of this function the properties will be outdated. Is this desirable?

Reply:
No, the `KeyStore` together with everything else inside a `StorageObject` are not meant to be used from multiple threads, strictly single threaded logic will be here. The parallelization will be done by using multiple shards per storage engine.

Reply:
:gem: completely forgot about that. Second time you mention this ;)